Panzer, Intrepid2: Compilation errors with cuda/12.4 + gcc/13.2.0

## Bug Report
@trilinos/panzer  @trilinos/intrepid2 

### Description
When testing a Cuda build with cuda/12.4 and gcc/13.2.0 I encountered compilation errors with Panzer, though they refer to Intrepid2 files:

```
FAILED: packages/panzer/disc-fe/src/CMakeFiles/panzer-disc-fe.dir/evaluators/Panzer_Dirichlet_Residual_EdgeBasis.cpp.o
...
/home/ndellin/Trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:580:1: error: conflicting declaration 'std::map<std::pair<__remove_cv(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >), __remove_cv(int)>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortCoeffData'
  580 |   OrientationTools<T>::ortCoeffData;
      | ^ ~~~~~~~~~~~~~~~~~~
/home/ndellin/Trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:421:205: note: previous declaration as 'std::map<std::pair<std::__cxx11::basic_string<char>, int>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortCoeffData'
  421 |     static std::map<std::pair<std::string,ordinal_type>,CoeffMatrixDataViewType> ortCoeffData;
      |                                                                                                                                                                                                             ^
/home/ndellin/Trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:584:1: error: conflicting declaration 'std::map<std::pair<__remove_cv(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >), __remove_cv(int)>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortInvCoeffData'
  584 |   OrientationTools<T>::ortInvCoeffData;
      | ^ ~~~~~~~~~~~~~~~~~~
/home/ndellin/Trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:422:205: note: previous declaration as 'std::map<std::pair<std::__cxx11::basic_string<char>, int>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortInvCoeffData'
  422 |     static std::map<std::pair<std::string,ordinal_type>,CoeffMatrixDataViewType> ortInvCoeffData;
```

### Steps to Reproduce
1. SHA1:  4c2c77595fa7c3c22dd609491e33d6b011692fb9
1. Configure script:  (sems modules)
```bash
export TRILINOS_DIR=<your-path>
export KOKKOS_DIR=$TRILINOS_DIR/packages/kokkos

module load sems-gcc/13.2.0 sems-cuda/12.4.0 sems-openblas/0.3.10 sems-openmpi/4.1.6 sems-cmake sems-ninja
module list
export OMPI_CXX=$KOKKOS_DIR/bin/nvcc_wrapper

export TPETRA_ASSUME_GPU_AWARE_MPI=0

# Configure Trilinos
cmake \
 -GNinja \
 -DCMAKE_INSTALL_PREFIX="${PWD}/install" \
 -DCMAKE_CXX_STANDARD="17" \
 -DTPL_ENABLE_MPI:BOOL=ON \
  -DTPL_ENABLE_CUDA=ON \
  -DTPL_ENABLE_BLAS:STRING=ON \
   -DBLAS_LIBRARY_DIRS:FILEPATH="$OPENBLAS_ROOT/lib" \
   -DBLAS_LIBRARY_NAMES:STRING="openblas" \
  -DTPL_ENABLE_LAPACK:STRING=ON \
   -DLAPACK_INCLUDE_DIRS:FILEPATH="$OPENBLAS_ROOT/include" \
   -DLAPACK_LIBRARY_DIRS:FILEPATH="$OPENBLAS_ROOT/lib" \
   -DLAPACK_LIBRARY_NAMES:STRING="openblas" \
  -DTPL_ENABLE_CUSPARSE:BOOL=ON \
  -DTPL_ENABLE_CUBLAS:BOOL=ON \
 -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
 -DTrilinos_ENABLE_TESTS=OFF \
 -DTrilinos_MUST_FIND_ALL_TPL_LIBS=TRUE \
 -DTrilinos_ENABLE_ALL_PACKAGES=OFF \
 -DTrilinos_ENABLE_COMPLEX=ON \
 -DTrilinos_ENABLE_OpenMP=OFF \
 -DTrilinos_ENABLE_Kokkos=ON \
   -D Kokkos_ENABLE_TESTS=OFF \
 -DKokkos_ENABLE_CUDA=ON \
 -DKokkos_ENABLE_CUDA_LAMBDA=ON \
 -DKokkos_ENABLE_CUDA_UVM=OFF \
  -DTrilinos_ENABLE_KokkosKernels=ON \
   -D KokkosKernels_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Tpetra=ON \
   -D Tpetra_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Sacado=ON \
   -D Sacado_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Stokhos=ON \
   -D Stokhos_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Adelus=ON \
   -D Adelus_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Compadre=ON \
   -D Compadre_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Amesos2=ON \
   -D Amesos2_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Zoltan2=ON \
   -D Zoltan2_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Ifpack2=ON \
   -D Ifpack2_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Belos=ON \
   -D Belos_ENABLE_TESTS=OFF \
  -DTrilinos_ENABLE_Intrepid2=ON \
   -D Intrepid2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Panzer=ON \
   -D Panzer_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_MueLu=ON \
   -D MueLu_ENABLE_TESTS=ON \
\
  -DTPL_ENABLE_Matio=OFF \
  -DTPL_ENABLE_Netcdf=OFF \
  -DTrilinos_ENABLE_ShyLU_NodeTacho=OFF \
  -DTrilinos_ENABLE_SEACAS=OFF \
$TRILINOS_DIR

make -j16
```

@CamelliaDPG I've also hit this error.  I'm using the same gcc and CUDA versions as @ndellingwood.   I tried commenting out https://github.com/trilinos/Trilinos/blob/b895aa6850943911c4b9388249a00a4d372f601e/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp#L545-L551

That results in the error:
```
/projects/sems/install/rhel8-x86_64/sems-compilers/binutils/2.42/gcc/8.3.0/bl7yrdz/bin/ld: packages/panzer/adapters-stk/src/libpanzer-stk.so.16.1.0: undefined reference to `Intrepid2::OrientationTools<Kokkos::Cuda>::ortCoeffData[abi:cxx11]'
/projects/sems/install/rhel8-x86_64/sems-compilers/binutils/2.42/gcc/8.3.0/bl7yrdz/bin/ld: packages/panzer/adapters-stk/src/libpanzer-stk.so.16.1.0: undefined reference to `Intrepid2::OrientationTools<Kokkos::Cuda>::ortInvCoeffData[abi:cxx11]'
collect2: error: ld returned 1 exit status
```

@mperego Perhaps I don't understand what you're asking for.  I see the error, for example, in this case:

```
/jhu/trilinos/Trilinos-trilinos/packages/panzer/disc-fe/src/Panzer_Interpolation.cpp
/ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:547:1: error: conflicting declaration 'std::map<std::pair<__remove_cv(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >), __remove_cv(int)>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortCoeffData'
  547 |   OrientationTools<T>::ortCoeffData;
      | ^ ~~~~~~~~~~~~~~~~~~
/ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:388:222: note: previous declaration as 'std::map<std::pair<std::__cxx11::basic_string<char>, int>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortCoeffData'
  388 |     static std::map<std::pair<std::string,ordinal_type>,CoeffMatrixDataViewType> ortCoeffData;
      |                                                                                                                                                                                                                              ^
/ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:551:1: error: conflicting declaration 'std::map<std::pair<__remove_cv(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >), __remove_cv(int)>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortInvCoeffData'
  551 |   OrientationTools<T>::ortInvCoeffData;
      | ^ ~~~~~~~~~~~~~~~~~~
/ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:389:222: note: previous declaration as 'std::map<std::pair<std::__cxx11::basic_string<char>, int>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortInvCoeffData'
  389 |     static std::map<std::pair<std::string,ordinal_type>,CoeffMatrixDataViewType> ortInvCoeffData;
```

After adding the code, I still see the same error.
```
es/panzer-disc-fe.dir/Panzer_Interpolation.cpp.o -c /ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/panzer/disc-fe/src/Panzer_Interpolation.cpp
/ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:554:1: error: conflicting declaration 'std::map<std::pair<__remove_cv(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >), __remove_cv(int)>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortCoeffData'
  554 |   OrientationTools<T>::ortCoeffData;
      | ^ ~~~~~~~~~~~~~~~~~~
/ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:395:222: note: previous declaration as 'std::map<std::pair<std::__cxx11::basic_string<char>, int>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortCoeffData'
  395 |     static std::map<std::pair<std::string,ordinal_type>,CoeffMatrixDataViewType> ortCoeffData;
      |                                                                                                                                                                                                                              ^
/ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:558:1: error: conflicting declaration 'std::map<std::pair<__remove_cv(const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >), __remove_cv(int)>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortInvCoeffData'
  558 |   OrientationTools<T>::ortInvCoeffData;
      | ^ ~~~~~~~~~~~~~~~~~~
/ascldap/users/jhu/trilinos/Trilinos-trilinos/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp:396:222: note: previous declaration as 'std::map<std::pair<std::__cxx11::basic_string<char>, int>, Kokkos::View<double****, DeviceType> > Intrepid2::OrientationTools<DeviceType>::ortInvCoeffData'
  396 |     static std::map<std::pair<std::string,ordinal_type>,CoeffMatrixDataViewType> ortInvCoeffData;
      |                                                                                                                                                                                                                              ^
```

--- a/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp
+++ b/packages/intrepid2/src/Orientation/Intrepid2_OrientationTools.hpp
@@ -380,13 +380,15 @@ namespace Intrepid2 {
 
     /** \brief  subcell ordinal, orientation, matrix m x n
      */
-    typedef Kokkos::View<double****,DeviceType> CoeffMatrixDataViewType;
+    using CoeffMatrixDataViewType = Kokkos::View<double****,DeviceType>;
 
     // 
     /** \brief  key :: basis name, order, value :: matrix data view type 
      */
-    static std::map<std::pair<std::string,ordinal_type>,CoeffMatrixDataViewType> ortCoeffData;
-    static std::map<std::pair<std::string,ordinal_type>,CoeffMatrixDataViewType> ortInvCoeffData;
+    using OrtCoeffDataType = std::map<std::pair<std::string,ordinal_type>,Kokkos::View<double****,DeviceType> >;
+
+    static OrtCoeffDataType ortCoeffData;
+    static OrtCoeffDataType ortInvCoeffData;
     
   private:
 
@@ -541,14 +543,13 @@ namespace Intrepid2 {
                               const BasisTypeLeft* basisLeft,
                               const BasisTypeRight* basisRight);
   };
-  
+
+  //Definition of static members
   template<typename T> 
-  std::map<std::pair<std::string,ordinal_type>, typename OrientationTools<T>::CoeffMatrixDataViewType>
-  OrientationTools<T>::ortCoeffData;
+  typename OrientationTools<T>::OrtCoeffDataType OrientationTools<T>::ortCoeffData;
 
   template<typename T> 
-  std::map<std::pair<std::string,ordinal_type>, typename OrientationTools<T>::CoeffMatrixDataViewType>
-  OrientationTools<T>::ortInvCoeffData;
+  typename OrientationTools<T>::OrtCoeffDataType OrientationTools<T>::ortInvCoeffData;
 }
 
 // include templated function definitions

